Mathlib/RingTheory/Valuation/ValExtension.lean

```diff
@@ -0,0 +1,181 @@
+    [LinearOrderedCommMonoidWithZero ΓR] [LinearOrderedCommMonoidWithZero ΓA]
+    [Algebra R A] (vR : Valuation R ΓR) (vA : Valuation A ΓA) [IsValExtension vR vA]
+
+-- @[simp] does not work because `vR` cannot be inferred from `R`.
+theorem val_map_le_iff (x y : R) : vA (algebraMap R A x) ≤ vA (algebraMap R A y) ↔ vR x ≤ vR y :=
+  (IsEquiv.val_le val_isEquiv_comap).symm
+
+theorem val_map_lt_iff (x y : R) : vA (algebraMap R A x) < vA (algebraMap R A y) ↔ vR x < vR y :=
+  (IsEquiv.val_lt val_isEquiv_comap).symm
+
+theorem val_map_eq_iff (x y : R) : vA (algebraMap R A x) = vA (algebraMap R A y) ↔ vR x = vR y :=
+  (IsEquiv.val_eq val_isEquiv_comap).symm
+
+theorem val_map_le_one_iff (x : R) : vA (algebraMap R A x) ≤ 1 ↔ vR x ≤ 1 :=
+  (IsEquiv.val_le_one val_isEquiv_comap).symm
+
+theorem val_map_lt_one_iff (x : R) : vA (algebraMap R A x) < 1 ↔ vR x < 1 :=
+  (IsEquiv.val_lt_one val_isEquiv_comap).symm
+
+theorem val_map_eq_one_iff (x : R) : vA (algebraMap R A x) = 1 ↔ vR x = 1 :=
+  (IsEquiv.val_eq_one val_isEquiv_comap).symm
+
+instance id : IsValExtension vR vR where
+  val_isEquiv_comap := by
+    simp only [Algebra.id.map_eq_id, comap_id]
+    rfl
+
+end CoeLemma
+
+section mk'
+
+/--
+When `K` is a field, if the preimage of the valuation integers of `A` equals to the valuation
+integers of `K`, then the valuation on `A` is an extension of valuation on `K`.
+-/
+theorem ofIntegerComap {K A ΓK ΓA : Type*} [Field K] [Ring A]
+    [LinearOrderedCommGroupWithZero ΓK] [LinearOrderedCommGroupWithZero ΓA]
+  [Algebra K A] (vK : Valuation K ΓK) (vA : Valuation A ΓA)
+    (h : vA.integer.comap (algebraMap K A) = vK.integer) : IsValExtension vK vA where
+  val_isEquiv_comap := by
+    rw [isEquiv_iff_val_le_one]
+    intro x
+    rw [← Valuation.mem_integer_iff, ← Valuation.mem_integer_iff]
+    rw [← h]
+    rfl
+
+end mk'
+
+section integer
+
+variable {R A ΓR ΓA : Type*} [CommRing R] [Ring A]
+    [LinearOrderedCommGroupWithZero ΓR] [LinearOrderedCommGroupWithZero ΓA]
+    [Algebra R A] {vR : Valuation R ΓR} {vA : Valuation A ΓA} [IsValExtension vR vA]
+
+instance integerAlgebra : Algebra vR.integer vA.integer where
+    smul r a := {
+      val := r • a,
+      property := by
+        rw [Valuation.mem_integer_iff,
+          show r • ↑a = algebraMap R A r * a by exact (Algebra.smul_def r (a : A))]
+        norm_num
+        apply mul_le_one'
+        · rw [val_map_le_one_iff vR]
+          exact r.2
+        · exact a.2
+    }
+    toFun r := {
+      val := algebraMap R A r,
+      property := by
+        simp only [Valuation.mem_integer_iff]
+        rw [val_map_le_one_iff vR]
+        exact r.2
+    }
+    map_one' := by
+      ext
+      simp
+    map_mul' _ _ := by
+      ext
+      simp
+    map_zero' := by
+      ext
+      simp
+    map_add' _ _ := by
+      ext
+      simp
+    commutes' _ _ := by
+      ext
+      exact Algebra.commutes _ _
+    smul_def' _ _ := by
+      ext
+      exact Algebra.smul_def _ _
```

Comment:
```suggestion
  smul r a := ⟨r • a,
    Algebra.smul_def r (a : A) ▸ mul_mem ((val_map_le_one_iff vR vA _).mpr r.2) a.2⟩
  __ := (algebraMap R A).restrict vR.integer vA.integer
    (by simp [Valuation.mem_integer_iff, val_map_le_one_iff vR vA])
  commutes' _ _ := Subtype.ext (Algebra.commutes _ _)
  smul_def' _ _ := Subtype.ext (Algebra.smul_def _ _)
```
Also please provide a simp lemma for smul.

Reply:
Thank you! I will add `coe_smul_integer`